### PR TITLE
Revert "Revert "Watcher implementation as publisher task""

### PR DESCRIPTION
**Pull request by akashgoyal96**
Reverts razorpay/metro#435

### CODE REVIEW

**Review by sakshi-goyal-razorpay — APPROVED** [review]

File: service/web/publisherserver.go [hunk 1]
```diff
@@ -29,16 +30,38 @@ func newPublisherServer(projectCore project.ICore, brokerStore brokerstore.IBrok
...
+		topicCacheCount.WithLabelValues(env, "Hit", req.Topic).Inc()
+	} else {
+		topicCacheCount.WithLabelValues(env, "Miss", req.Topic).Inc()
+		logger.Ctx(ctx).Infow(
+			"PublishServer: Topic doesn't exist inside the cache..",
+			"req",
+			req.Topic)
+
+		// Fallback method to check topic existence in the Registry
+		if ok, err := s.topicCore.ExistsWithName(
```
Comment (vnktram) [inline service/web/publisherserver.go:54]: If topic is not found in cache but is present in the registry then we should update the local cache data with the topic info.

File: internal/tasks/publishertask.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+		return terr
+	}
+
+	topicData := make(map[string]bool)
+	for _, topic := range topics {
+		topicData[topic.Name] = true
+	}
+
+	TopicCacheData = topicData
```
Comment (vnktram) [inline internal/tasks/publishertask.go:150]: Wrap this around a semaphore.

File: internal/tasks/publishertask.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+	if _, ok := TopicCacheData[topic]; ok {
+
+		return true
+	}
+
+	return false
+}
+
+// UpdateTopicCache is to update the cache
+func UpdateTopicCache(ctx context.Context, topic string) {
```
Comment (sakshi-goyal-razorpay) [inline internal/tasks/publishertask.go:167]: you can have a flag in this function that defines whether entire cache needs to be updated or just for topic. 
and based on that, you can update the topicCacheData inside a semaphore.
You can define a struct here, to handle both cases.

File: internal/tasks/publishertask.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+		return true
+	}
+
+	return false
+}
+
+// UpdateTopicCache is to update the cache
+func UpdateTopicCache(ctx context.Context, topicMap map[string]bool, specificTopicBool bool) {
+	sem <- 1
```
Comment (vnktram) [inline internal/tasks/publishertask.go:170]: Let's use a simple Mutex lock and avoid re-inventing the wheel here. https://pkg.go.dev/sync#Mutex.Lock

File: internal/tasks/publishertask.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+	if _, ok := TopicCacheData[topic]; ok {
+
+		return true
+	}
+
+	return false
+}
+
+// UpdateTopicCache is to update the cache
+func UpdateTopicCache(ctx context.Context, topicMap map[string]bool, specificTopicBool bool) {
```
Comment (vnktram) [inline internal/tasks/publishertask.go:169]: `specificTopicBool` needs better documentation. If possible avoid passing a bool altogether and find a better way to update the cache.

**Review by vnktram — CHANGES_REQUESTED** [review]

File: internal/tasks/publishertask.go (hunk 5 above)
Comment (akashgoyal96) [reply internal/tasks/publishertask.go:171]: I have added detailed comment for `specificTopicBool `

File: internal/tasks/publishertask.go (hunk 4 above)
Comment (vnktram) [reply internal/tasks/publishertask.go]: Why was this marked as resolved? Can we use a mutex instead?

**Review by vnktram — APPROVED** [review]

**Pull request merged**